README.md
# just_audio_cache

[![pub package](https://img.shields.io/pub/v/just_audio_cache.svg)](https://pub.dev/packages/just_audio_cache)

Collection of extension function of [just_audio](https://pub.dev/packages/just_audio) package for auto-handle caching audio files

## How to use

<p> The premise is you already have just_audio package in your dependency </p>

Core method `dynamicSet` would auto handle caching, if the target file not in cache, would download it and push in cache.

The next time the same url request call, would call the local file instead of requesting.

`url` is your audio source, a unique key that represents the stored file path,

`pushIfNotExisted` if true, when the file not exists, would download the file and push in cache

`excludeCallback` a callback function where you can specify which file you don't want to be cached
```dart
import 'package:just_audio_cache/just_audio_cache.dart';

void _play() async {
   // dynamic set your audio source
   await _player.dynamicSet(your_url);
   _player.play();
}
```

<p> If you want to check if the target file exist in cache </p>

```dart
_player.existedInLocal(your_url);
```

<p> Manually add file to cache </p>

```dart
_player.cacheFile(your_audio_url);
```

<p> Get Cached file for the following url </p>

```dart
_player.getCachedFile(url: your_url);
```

<p> Clear all the cache from storage </p>

``` dart
_player.clear();
```








CHANGELOG.md
## 0.1.0

Official version

### 0.1.1
Add [dynamicSetAll] method
* Store all the playlist file async with just 1 method


test/just_audio_cache_test.dart
import 'dart:io';

import 'package:flutter/services.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:just_audio/just_audio.dart';
import 'package:just_audio_cache/just_audio_cache.dart';
import 'package:path_provider/path_provider.dart';

void main() {
  TestWidgetsFlutterBinding.ensureInitialized();
  group('audio_test', () {
    test('download_test', () async {
      final spMockCacheStrs = <String, String>{};
      final directory = await Directory.systemTemp.createTemp();
      // Mock out the MethodChannel for the path_provider plugin.
      const MethodChannel('plugins.flutter.io/path_provider')
          .setMockMethodCallHandler((MethodCall methodCall) async {
        // If you're getting the apps documents directory, return the path to the
        // temp directory on the test environment instead.
        if (methodCall.method == 'getApplicationDocumentsDirectory') {
          return directory.path;
        }
        if (methodCall.method == 'getTemporaryDirectory') {
          return directory.path;
        }
        return null;
      });
      // Mock out the MethodChannel for the share_preference plugin.
      const MethodChannel('plugins.flutter.io/shared_preferences')
          .setMockMethodCallHandler((MethodCall methodCall) async {
        // If you're getting the apps documents directory, return the path to the
        // temp directory on the test environment instead.
        if (methodCall.method == 'getAll') {
          return spMockCacheStrs;
        }
        if (methodCall.method == 'setString') {
          spMockCacheStrs[methodCall.arguments] = methodCall.arguments;
          return true;
        }
        if (methodCall.method == 'getString') {
          return spMockCacheStrs[methodCall.arguments];
        }
      });

      final dir = await getApplicationDocumentsDirectory();
      final player = AudioPlayer();
      final sources = [
        'https://www.mboxdrive.com/m54.mp3',
        'https://www.mboxdrive.com/%c3%86STRAL%20-%20half%20light.mp3',
        'https://www.mboxdrive.com/%ca%8f%e1%b4%80%20%ca%99%e1%b4%9c%e1%b4%85%e1%b4%9c.mp3'
      ];
      await player.dynamicSetAll(sources);
      // await player.dynamicSet(
      //     url: 'https://www.mboxdrive.com/m54.mp3',
      //     path: dir.path,
      //     excludeCallback: (_) => true);
      for (final url in sources) {
        expect(await player.existedInLocal(url: url), true);
      }

      /*test('download_and_cache_file', () async {
      final player = AudioPlayer();
      player.cacheFile(url: 'https://www.mboxdrive.com/m54.mp3');
    });

    test('playFromFile', () async {
      final player = AudioPlayer();
      final dir = await getApplicationDocumentsDirectory();
      final url = 'https://www.mboxdrive.com/m54.mp3';
      if (await player.existedInLocal(url: url)) {
        expect(player.getCachedPath(url: url),
            dir.path + '/audio_cache/' + 'm54.mp3');
      }
    });*/
    });
  });
}


lib/src/audio_player_utils.dart
import 'dart:io';

import 'package:just_audio/just_audio.dart';
import 'package:just_audio_cache/src/io_client.dart';
import 'package:just_audio_cache/src/parser.dart';
import 'package:path_provider/path_provider.dart';
import 'package:shared_preferences/shared_preferences.dart';

extension AudioPlayerExtension on AudioPlayer {
  static SharedPreferences? _sp;

  /// Check if the file corresponds the url exists in local
  Future<bool> existedInLocal({required String url}) async {
    if (_sp == null) _sp = await SharedPreferences.getInstance();

    return _sp!.getString(url) != null;
  }

  /// Get audio file cache path
  Future<String?> getCachedPath({required String url}) async {
    if (_sp == null) _sp = await SharedPreferences.getInstance();

    return _sp!.getString(getUrlSuffix(url));
  }

  /// Get audio from local if exist, otherwise download from network
  /// [url] is your audio source, a unique key that represents the stored file path,
  /// [path] the storage path you want to save your cache
  /// [pushIfNotExisted] if true, when the file not exists, would download the file and push in cache
  /// [excludeCallback] a callback function where you can specify which file you don't want to be cached,
  ///  (return `true` if we want to exclude the specific source)
  Future<Duration?> dynamicSet({
    required String url,
    String? path,
    bool pushIfNotExisted = true,
    bool excludeCallback(url)?,
    bool preload = true,
  }) async {
    if (_sp == null) _sp = await SharedPreferences.getInstance();

    if (excludeCallback != null) {
      pushIfNotExisted = excludeCallback(url);
    }

    final dirPath = path ?? (await _openDir()).path;
    // File check
    if (await existedInLocal(url: url)) {
      // existed, play from local file
      try {
        return await setFilePath(_sp!.getString(url)!, preload: preload);
      } catch (e) {
        print(e);
      }
    }

    final duration = await setUrl(url, preload: preload);

    // download to cache after setUrl in order to show the audio buffer state
    if (pushIfNotExisted) {
      final key = getUrlSuffix(url);
      IoClient.download(url: url, path: dirPath + '/' + key).then((storedPath) {
        if (storedPath != null) {
          _sp!.setString(url, storedPath);
        }
      });
    }

    return duration;
  }

  /// Cache a collection of audio source
  /// [sources] target sources for your playlist
  /// [path] The dir path where sources store
  /// [excluded] the sources you don't want to save in storage
  Future<List<Duration?>> dynamicSetAll(
    List<String> sources, [
    String? path,
    List<String>? excluded,
  ]) async {
    final durations = const <Duration?>[];
    for (final url in sources) {
      durations.add(
        await dynamicSet(
          url: url,
          path: path,
          excludeCallback: (url) => excluded != null && excluded.contains(url),
        ),
      );
    }
    return durations;
  }

  Future<void> cacheFile({required String url, String? path}) async {
    final dirPath = path ?? (await _openDir()).path;
    final storedPath = await IoClient.download(url: url, path: dirPath);
    if (storedPath != null) {
      _sp!.setString(url, storedPath);
    }
  }

  /// Clear all the cache in the app dir
  Future<void> clearCache({String? path}) async {
    final dir = path != null ? Directory(path) : (await _openDir());
    return dir.deleteSync();
  }

  Future<void> playFromFile({required String filePath}) async {
    await setFilePath(filePath);
    return await play();
  }

  /*Future<bool> _isKeyExisted(String key) async {
    if (_sp == null) _sp = await SharedPreferences.getInstance();

    return _sp!.getString(key) != null;
  }*/

  Future<Directory> _openDir() async {
    final dir = await getApplicationDocumentsDirectory();
    final Directory targetDir = Directory(dir.path + '/audio_cache');
    if (!targetDir.existsSync()) {
      targetDir.createSync();
    }
    return targetDir;
  }
}


example/lib/main.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:just_audio/just_audio.dart';
import 'package:just_audio_cache/just_audio_cache.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatefulWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  late final AudioPlayer _player;
  final url = 'https://www.mboxdrive.com/m54.mp3';
  final sources = [
    'https://www.mboxdrive.com/m54.mp3',
    'https://www.mboxdrive.com/%c3%86STRAL%20-%20half%20light.mp3',
    'https://www.mboxdrive.com/%ca%8f%e1%b4%80%20%ca%99%e1%b4%9c%e1%b4%85%e1%b4%9c.mp3'
  ];
  PlayerState? _state;

  @override
  void initState() {
    _player = AudioPlayer();
    //_player.dynamicSet(url: url);
    _player.dynamicSetAll(sources);
    _player.playerStateStream.listen((state) {
      setState(() {
        _state = state;
      });
      print(state);
    });
    super.initState();
  }

  void _playAudio() async {
    _player.play();
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Container(
        child: Center(
          child: _audioStateWidget(),
        ),
      ),
    );
  }

  Widget _audioStateWidget() {
    if (_state == null) return _playButton;

    if (_state!.playing) {
      return _pauseButton;
    } else {
      return _playButton;
    }
  }

  Widget get _pauseButton => ElevatedButton(
        onPressed: () {
          _player.pause();
        },
        child: Text('Pause'),
      );

  Widget get _playButton => ElevatedButton(
        onPressed: () {
          _playAudio();
        },
        child: Text('Play'),
      );
}
